### PR TITLE
Add SparkFun Qwiic Button binding

### PR DESCRIPTION
**Pull request by Esbob**
@Ellerbach and @krwq, this PR is the same as PR #1204 that got closed when I renamed from master to main.
I also removed the third license line. All comments should be fixed now.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/iot/pull/1427)

### CODE REVIEW

File: src/devices/QwiicButton/FlagsHelper.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+namespace Iot.Device.QwiicButton
```
Comment (Ellerbach) [inline src/devices/QwiicButton/FlagsHelper.cs:6]: as far as I remember @pgrawehr suggested to move this file to src/devices/common
I do believe it makes sense. It's a handy helpers that can benefit to more internal project.
@krwq any feedback?

**Review by Ellerbach — APPROVED** [review]
Looks good to me. Thanks for moving this forward.

File: src/devices/QwiicButton/tests/QwiicButton.Tests.csproj [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <!--
+  // NOTE: THIS TEST PROJECT IS NOT REQUIRED AND CAN BE DELETED IF NOT USED.
+  -->
+  
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
```
Comment (Ellerbach) [inline src/devices/QwiicButton/tests/QwiicButton.Tests.csproj:8]: should the tests run on net5 or fine with core 2.1?

File: src/devices/QwiicButton/samples/QwiicButton.Samples.csproj [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Comment (Ellerbach) [inline src/devices/QwiicButton/samples/QwiicButton.Samples.csproj:5]: the sample should target `net5.0`

File: src/devices/QwiicButton/QwiicButton.csproj [hunk 4]
```diff
@@ -0,0 +1,16 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
```
Comment (Ellerbach) [inline src/devices/QwiicButton/QwiicButton.csproj:4]: Target framework should be: `<TargetFrameworks>net5.0;netcoreapp2.1</TargetFrameworks>`

File: src/devices/QwiicButton/tests/QwiicButton.Tests.csproj (hunk 2 above)
Comment (Esbob) [reply src/devices/QwiicButton/tests/QwiicButton.Tests.csproj]: Changed to net5.0

File: src/devices/QwiicButton/samples/QwiicButton.Samples.csproj (hunk 3 above)
Comment (Esbob) [reply src/devices/QwiicButton/samples/QwiicButton.Samples.csproj]: Fixed

File: src/devices/QwiicButton/QwiicButton.csproj (hunk 4 above)
Comment (Esbob) [reply src/devices/QwiicButton/QwiicButton.csproj]: Fixed

File: src/devices/QwiicButton/FlagsHelper.cs (hunk 1 above)
Comment (krwq) [reply src/devices/QwiicButton/FlagsHelper.cs:6]: I'm ok either way, we can suggest this change next time we notice someone doing something similar

File: src/devices/QwiicButton/FlagsHelper.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+                Set(ref flags, flag);
+            }
+            else
+            {
+                Unset(ref flags, flag);
+            }
+        }
+
+        private static void Set<T>(ref T flags, T flag)
+            where T : struct
```
Comment (krwq) [inline src/devices/QwiicButton/FlagsHelper.cs:48]: why isn't T just constrained to `enum`? Can you avoid casting to object below if you do that?

File: src/devices/QwiicButton/I2cRegisterAccess.cs [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    /// it often means that you have to read the data at several individual 8-bit addresses and
+    /// assemble this data in the correct order to extract the corresponding 16/32/64-bit word.
+    /// For instance, for some devices a register address such as 0x1234 means that you need to
+    /// read/write from registers 0x12 and 0x34. Both registers will hold 8-bits of information,
+    /// which together form the actual 16-bit word referenced by the hexadecimal 0x1234.
+    /// Read through the data sheets/manuals for your specific I2C device for more information
+    /// on its register addressing.
+    /// </remarks>
+    /// </summary>
+    public sealed class I2cRegisterAccess : IDisposable
```
Comment (krwq) [inline src/devices/QwiicButton/I2cRegisterAccess.cs:25]: this shouldn't be public, it's an internal utility

File: src/devices/QwiicButton/I2cRegisterAccessOfT.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    /// it often means that you have to read the data at several individual 8-bit addresses and
+    /// assemble this data in the correct order to extract the corresponding 16/32/64-bit word.
+    /// For instance, for some devices a register address such as 0x1234 means that you need to
+    /// read/write from registers 0x12 and 0x34. Both registers will hold 8-bits of information,
+    /// which together form the actual 16-bit word referenced by the hexadecimal 0x1234.
+    /// Read through the data sheets/manuals for your specific I2C device for more information
+    /// on its register addressing.
+    /// </remarks>
+    /// </summary>
+    public sealed class I2cRegisterAccess<TRegisterMap> : IDisposable
```
Comment (krwq) [inline src/devices/QwiicButton/I2cRegisterAccessOfT.cs:37]: this similarly, it should be internal

File: src/devices/QwiicButton/I2cRegisterAccess.cs (hunk 6 above)
Comment (krwq) [reply src/devices/QwiicButton/I2cRegisterAccess.cs:25]: (it's only useful for implementing devices, not for anyone outside of repo, they should use devices directly)

File: src/devices/QwiicButton/FlagsHelper.cs (hunk 5 above)
Comment (Esbob) [reply src/devices/QwiicButton/FlagsHelper.cs:48]: I don't remember the reason now, but it doesn't work when constrained to an enum.

File: src/devices/QwiicButton/I2cRegisterAccess.cs (hunk 6 above)
Comment (Esbob) [reply src/devices/QwiicButton/I2cRegisterAccess.cs]: Fixed

File: src/devices/QwiicButton/I2cRegisterAccessOfT.cs (hunk 7 above)
Comment (Esbob) [reply src/devices/QwiicButton/I2cRegisterAccessOfT.cs]: Fixed

**Comment by Esbob** [comment]
@krwq, can you please approve this PR, so we can finally close it?

**Pull request closed without merging**